<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="$(arg robot_name)">
  <xacro:include filename="$(find sr_moveit_hand_config)/config/shadowhand.srdf.xacro"/>
  <xacro:arg name="prefix" default="rh_"/>
  <xacro:arg name="ff" default="1"/>
  <xacro:arg name="mf" default="1"/>
  <xacro:arg name="rf" default="1"/>
  <xacro:arg name="lf" default="1"/>
  <xacro:arg name="th" default="1"/>
  <xacro:arg name="ff_bio" default="0"/>
  <xacro:arg name="mf_bio" default="0"/>
  <xacro:arg name="rf_bio" default="0"/>
  <xacro:arg name="lf_bio" default="0"/>
  <xacro:arg name="th_bio" default="0"/>     
  <xacro:arg name="is_lite" default="0"/>
  <xacro:arg name="robot_name" default="shadowhand_motor"/>
  
  <!-- hand group --> 
  <group name="shadow_hand">
    <xacro:if value="$(arg th)">
      <xacro:if value="$(arg th_bio)">
        <xacro:property name="distal" value="biotac"/>
      </xacro:if>
      <xacro:unless value="$(arg th_bio)">
        <xacro:property name="distal" value="distal"/>
      </xacro:unless>
      <xacro:thumblinks prefix="" />
      <xacro:thumbjoints prefix=""/>
    </xacro:if>
    <xacro:if value="$(arg ff)">
      <xacro:if value="$(arg ff_bio)">
        <xacro:property name="distal" value="biotac"/>
        <xacro:property name="middle" value="adapter"/>
      </xacro:if>
      <xacro:unless value="$(arg ff_bio)">  
        <xacro:property name="distal" value="distal"/>
        <xacro:property name="middle" value="middle"/>
      </xacro:unless>
      <xacro:fingerlinks prefix="" link_prefix="ff" lf="0"/>
      <xacro:fingerjoints prefix="" joint_prefix="FF" lf="0"/>
    </xacro:if>
    <xacro:if value="$(arg mf)">
      <xacro:if value="$(arg mf_bio)">
        <xacro:property name="distal" value="biotac"/>
        <xacro:property name="middle" value="adapter"/>
      </xacro:if>
      <xacro:unless value="$(arg mf_bio)">  
        <xacro:property name="distal" value="distal"/>
        <xacro:property name="middle" value="middle"/>
      </xacro:unless>
      <xacro:fingerlinks prefix="" link_prefix="mf" lf="0"/>
      <xacro:fingerjoints prefix="" joint_prefix="MF" lf="0"/>
    </xacro:if>
    <xacro:if value="$(arg rf)">
      <xacro:if value="$(arg rf_bio)">
        <xacro:property name="distal" value="biotac"/>
        <xacro:property name="middle" value="adapter"/>
      </xacro:if>
      <xacro:unless value="$(arg rf_bio)">  
        <xacro:property name="distal" value="distal"/>
        <xacro:property name="middle" value="middle"/>
      </xacro:unless>
      <xacro:fingerlinks prefix="" link_prefix="rf" lf="0"/>
      <xacro:fingerjoints prefix="" joint_prefix="RF" lf="0"/>
    </xacro:if>
    <xacro:unless value="$(arg is_lite)">
      <xacro:if value="$(arg lf)">
        <xacro:if value="$(arg lf_bio)">
          <xacro:property name="distal" value="biotac"/>
          <xacro:property name="middle" value="adapter"/>
        </xacro:if>
        <xacro:unless value="$(arg lf_bio)">  
          <xacro:property name="distal" value="distal"/>
          <xacro:property name="middle" value="middle"/>
        </xacro:unless>
        <xacro:fingerlinks prefix="" link_prefix="lf" lf="1"/>
        <xacro:fingerjoints prefix="" joint_prefix="LF" lf="1"/>
      </xacro:if>
      <joint name="WRJ2" />
      <link name="wrist" />
      <joint name="WRJ1" />
      <chain base_link="forearm" tip_link="palm" />
    </xacro:unless>
    <link name="palm" />
    <link name="forearm" />
    
    
    <group name="fingers" />
  </group>
 
 <!-- single finger groups -->  
  <xacro:if value="$(arg ff)">  
    <group name="first_finger">
      <xacro:fingerjoints prefix="" joint_prefix="FF" lf="0"/>
      <chain base_link="palm" tip_link="fftip" />
    </group>
  </xacro:if>
  <xacro:if value="$(arg mf)">
    <group name="middle_finger">
      <xacro:fingerjoints prefix="" joint_prefix="MF" lf="0"/>
      <chain base_link="palm" tip_link="mftip" />
    </group>
  </xacro:if>
  <xacro:if value="$(arg rf)">
    <group name="ring_finger">
      <xacro:fingerjoints prefix="" joint_prefix="RF" lf="0"/>
      <chain base_link="palm" tip_link="rftip" />
    </group>
  </xacro:if>
  <xacro:unless value="$(arg is_lite)">
    <xacro:if value="$(arg lf)">
      <group name="little_finger">
        <xacro:fingerjoints prefix="" joint_prefix="LF" lf="1"/>
        <chain base_link="palm" tip_link="lftip" />
      </group>
    </xacro:if>
  </xacro:unless>
  <xacro:if value="$(arg th)">
    <group name="thumb">
      <xacro:thumbjoints prefix=""/>
      <chain base_link="palm" tip_link="thtip" />
    </group>
  </xacro:if>
  
  <!-- all fingerstips --> 
  <group name="fingertips">
    <xacro:if value="$(arg ff)">
      <link name="fftip" />
    </xacro:if>
    <xacro:if value="$(arg mf)">
      <link name="mftip" />
    </xacro:if>
    <xacro:if value="$(arg rf)">
      <link name="rftip" />
    </xacro:if>
    <xacro:if value="$(arg lf)">
      <link name="lftip" />
    </xacro:if>
    <xacro:if value="$(arg th)">
      <link name="thtip" />
    </xacro:if>
  </group>
  
  <!-- all fingers group -->     
  <group name="fingers">
   <xacro:if value="$(arg th)">
      <xacro:thumbjoints prefix="" />
      <group name="thumb" />
    </xacro:if>
    <xacro:if value="$(arg ff)">
      <xacro:fingerjoints prefix="" joint_prefix="FF" lf="0"/>
      <group name="first_finger" />
    </xacro:if>
    <xacro:if value="$(arg mf)">
      <xacro:fingerjoints prefix="" joint_prefix="MF" lf="0"/>
      <group name="middle_finger" />
    </xacro:if>
    <xacro:if value="$(arg rf)">
      <xacro:fingerjoints prefix="" joint_prefix="RF" lf="0"/>
      <group name="ring_finger" />
    </xacro:if>
    <xacro:unless value="$(arg is_lite)">
      <xacro:if value="$(arg lf)">
        <xacro:fingerjoints prefix="" joint_prefix="LF" lf="1"/>
        <group name="little_finger" />
      </xacro:if>
    </xacro:unless>      
  </group>
  
   <!-- pairs with thumb for groups -->
  <xacro:if value="$(arg th)">
    <xacro:if value="$(arg ff)">
      <group name="first_finger_thumb">
        <xacro:thumbjoints prefix="" />
        <xacro:fingerjoints prefix="" joint_prefix="FF" lf="0"/>
      </group>
    </xacro:if>
    <xacro:if value="$(arg mf)">
      <group name="middle_finger_thumb">
        <xacro:thumbjoints prefix=""/>
        <xacro:fingerjoints prefix="" joint_prefix="MF" lf="0"/>
      </group>
    </xacro:if>
    <xacro:if value="$(arg rf)">
      <group name="ring_finger_thumb">
        <xacro:thumbjoints prefix=""/>
        <xacro:fingerjoints prefix="" joint_prefix="RF" lf="0"/>
      </group>
    </xacro:if>
    <xacro:unless value="$(arg is_lite)">
      <xacro:if value="$(arg lf)">
        <group name="little_finger_thumb">
          <xacro:thumbjoints prefix=""/>
          <xacro:fingerjoints prefix="" joint_prefix="LF" lf="1"/>
        </group>
      </xacro:if>
    </xacro:unless>  
  </xacro:if>   
    
  <!--GROUP STATES: Purpose: Define a named state for a particular group, in terms of joint values. This is useful to define states like 'folded arms'-->
  
  <!-- hand state --> 
  <group_state name="hand_zero" group="shadow_hand">
    <xacro:if value="$(arg th)">
        <xacro:thumb_state_open  prefix="" />
    </xacro:if>
    <xacro:if value="$(arg ff)">
        <xacro:finger_state_open  prefix="" joint_prefix="FF" lf="0" />
    </xacro:if>
    <xacro:if value="$(arg mf)">
        <xacro:finger_state_open  prefix="" joint_prefix="MF" lf="0" />
    </xacro:if>
    <xacro:if value="$(arg rf)">
        <xacro:finger_state_open  prefix="" joint_prefix="RF" lf="0" />
    </xacro:if>
    <xacro:unless value="$(arg is_lite)">
      <xacro:if value="$(arg lf)">
          <xacro:finger_state_open  prefix="" joint_prefix="LF" lf="1" />
      </xacro:if>
      <joint name="WRJ1" value="0" />
      <joint name="WRJ2" value="0" />
    </xacro:unless>
  </group_state>
  
  <!-- single fingers state --> 
  <xacro:if value="$(arg ff)">
    <group_state name="first_finger_open" group="first_finger">
        <xacro:finger_state_open  prefix="" joint_prefix="FF" lf="0" />
      </group_state>
  </xacro:if>
  <xacro:if value="$(arg mf)">
    <group_state name="middle_finger_open" group="middle_finger">
      <xacro:finger_state_open  prefix="" joint_prefix="MF" lf="0" />
    </group_state>
  </xacro:if>
  <xacro:if value="$(arg rf)">
    <group_state name="ring_finger_open" group="ring_finger">
      <xacro:finger_state_open  prefix="" joint_prefix="RF" lf="0" />
    </group_state>
  </xacro:if>
  <xacro:unless value="$(arg is_lite)">
    <xacro:if value="$(arg lf)">
      <group_state name="little_finger_open" group="little_finger">
        <xacro:finger_state_open  prefix="" joint_prefix="LF" lf="1" />
      </group_state>
    </xacro:if>
  </xacro:unless>
  <xacro:if value="$(arg th)">
    <group_state name="thumb_open" group="thumb">
        <xacro:thumb_state_open  prefix="" />
    </group_state>
  </xacro:if>

  <!-- all fingers state -->  
  <group_state name="fingers_open" group="fingers">
    <xacro:if value="$(arg th)">
        <xacro:thumb_state_open  prefix="" />
    </xacro:if>
    <xacro:if value="$(arg ff)">
        <xacro:finger_state_open  prefix="" joint_prefix="FF" lf="0" />
    </xacro:if>
    <xacro:if value="$(arg mf)">
        <xacro:finger_state_open  prefix="" joint_prefix="MF" lf="0" />
    </xacro:if>
    <xacro:if value="$(arg rf)">
        <xacro:finger_state_open  prefix="" joint_prefix="RF" lf="0" />
    </xacro:if>
    <xacro:unless value="$(arg is_lite)">
      <xacro:if value="$(arg lf)">
          <xacro:finger_state_open  prefix="" joint_prefix="LF" lf="1" />
      </xacro:if>
    </xacro:unless>
  </group_state>
 
  <!-- pairs with thumb for state -->
  <xacro:if value="$(arg th)">
    <xacro:if value="$(arg ff)">
      <xacro:finger_thumb_state_open  prefix="" finger_name="first_finger" joint_prefix="FF" lf="0"  is_lite="$(arg is_lite)" />
    </xacro:if>
    <xacro:if value="$(arg mf)">
      <xacro:finger_thumb_state_open  prefix="" finger_name="middle_finger" joint_prefix="MF" lf="0"  is_lite="$(arg is_lite)" />
    </xacro:if>
    <xacro:if value="$(arg rf)">
      <xacro:finger_thumb_state_open  prefix="" finger_name="ring_finger" joint_prefix="RF" lf="0"  is_lite="$(arg is_lite)" />
    </xacro:if>
    <xacro:unless value="$(arg is_lite)">
      <xacro:if value="$(arg lf)">
          <xacro:finger_thumb_state_open  prefix="" finger_name="little_finger" joint_prefix="LF" lf="1"  is_lite="$(arg is_lite)" />
      </xacro:if>
    </xacro:unless>
  </xacro:if>

  <!--END EFFECTOR: Purpose: Represent information about an end effector.-->
    
  <!-- single finger end-effectors --> 
  <xacro:if value="$(arg ff)">
    <xacro:finger_ee prefix="" link_prefix="ff" finger_name="first_finger" />
  </xacro:if>
  <xacro:if value="$(arg mf)">
    <xacro:finger_ee prefix="" link_prefix="mf" finger_name="middle_finger" />
  </xacro:if>
  <xacro:if value="$(arg rf)">
    <xacro:finger_ee prefix="" link_prefix="rf" finger_name="ring_finger" />
  </xacro:if>
  <xacro:unless value="$(arg is_lite)">
    <xacro:if value="$(arg lf)">
      <xacro:finger_ee prefix="" link_prefix="lf" finger_name="little_finger" />
    </xacro:if>
  </xacro:unless>  
  <xacro:if value="$(arg th)">
    <xacro:finger_ee prefix="" link_prefix="th" finger_name="thumb" />
  </xacro:if>
  
  <xacro:unless value="$(arg is_lite)">
    <end_effector name="shadow_hand_ee" parent_link="wrist" group="shadow_hand" />
  </xacro:unless>  
    
    
  <xacro:virtual_joints prefix="" />
    

  <!--DISABLE COLLISIONS: By default it is assumed that any link of the robot could potentially come into collision with any other link in the robot. This tag disables collision checking between a specified pair of links. -->

  <xacro:hand_collisions prefix="" /> 
  <!-- single fingers collision --> 
  <xacro:if value="$(arg ff)">
    <xacro:if value="$(arg ff_bio)">
      <xacro:property name="distal" value="biotac"/>
      <xacro:property name="middle" value="adapter"/>
    </xacro:if>
    <xacro:unless value="$(arg ff_bio)">
      <xacro:property name="distal" value="distal"/>
      <xacro:property name="middle" value="middle"/>
    </xacro:unless>
    <xacro:finger_collisions prefix="" link_prefix="ff" lf="0" />
  </xacro:if>
  <xacro:if value="$(arg mf)">
    <xacro:if value="$(arg mf_bio)">
      <xacro:property name="distal" value="biotac"/>
      <xacro:property name="middle" value="adapter"/>
    </xacro:if>
    <xacro:unless value="$(arg mf_bio)">
      <xacro:property name="distal" value="distal"/>
      <xacro:property name="middle" value="middle"/>
    </xacro:unless>
    <xacro:finger_collisions prefix="" link_prefix="mf" lf="0" />
  </xacro:if>
  <xacro:if value="$(arg rf)">
    <xacro:if value="$(arg rf_bio)">
      <xacro:property name="distal" value="biotac"/>
      <xacro:property name="middle" value="adapter"/>
    </xacro:if>
    <xacro:unless value="$(arg rf_bio)">
      <xacro:property name="distal" value="distal"/>
      <xacro:property name="middle" value="middle"/>
    </xacro:unless>
    <xacro:finger_collisions prefix="" link_prefix="rf" lf="0" />
  </xacro:if>
  <xacro:unless value="$(arg is_lite)">
    <xacro:if value="$(arg lf)">
      <xacro:if value="$(arg lf_bio)">
        <xacro:property name="distal" value="biotac"/>
        <xacro:property name="middle" value="adapter"/>
      </xacro:if>
      <xacro:unless value="$(arg lf_bio)">
        <xacro:property name="distal" value="distal"/>
        <xacro:property name="middle" value="middle"/>
      </xacro:unless>
       <xacro:finger_collisions prefix="" link_prefix="lf" lf="1" />
    </xacro:if>
  </xacro:unless>  
  <xacro:if value="$(arg th)">
    <xacro:if value="$(arg th_bio)">
      <xacro:property name="distal" value="biotac"/>
    </xacro:if>
    <xacro:unless value="$(arg th_bio)">
      <xacro:property name="distal" value="distal"/>
    </xacro:unless>
    <xacro:thumb_collisions prefix="" /> 
  </xacro:if>

  <!-- pairs with thumb for collision-->
  <xacro:if value="$(arg th)">
    <xacro:if value="$(arg th_bio)">
      <xacro:property name="thdistal" value="thbiotac"/>
    </xacro:if>
    <xacro:unless value="$(arg th_bio)">  
      <xacro:property name="thdistal" value="thdistal"/>
    </xacro:unless>
    <xacro:if value="$(arg ff)">
      <xacro:if value="$(arg ff_bio)">
        <xacro:property name="distal" value="biotac"/>
        <xacro:property name="middle" value="adapter"/>
      </xacro:if>
      <xacro:unless value="$(arg ff_bio)">  
        <xacro:property name="distal" value="distal"/>
        <xacro:property name="middle" value="middle"/>
      </xacro:unless>
      <xacro:finger_thumb_collisions  prefix="" link_prefix="ff" far="0" lf="0" />
    </xacro:if>
    <xacro:if value="$(arg mf)">
      <xacro:if value="$(arg mf_bio)">
        <xacro:property name="distal" value="biotac"/>
        <xacro:property name="middle" value="adapter"/>
      </xacro:if>
      <xacro:unless value="$(arg mf_bio)">  
        <xacro:property name="distal" value="distal"/>
        <xacro:property name="middle" value="middle"/>
      </xacro:unless>
      <xacro:finger_thumb_collisions  prefix="" link_prefix="mf" far="1" lf="0" />
    </xacro:if>
    <xacro:if value="$(arg rf)">
      <xacro:if value="$(arg rf_bio)">
        <xacro:property name="distal" value="biotac"/>
        <xacro:property name="middle" value="adapter"/>
      </xacro:if>
      <xacro:unless value="$(arg rf_bio)">  
        <xacro:property name="distal" value="distal"/>
        <xacro:property name="middle" value="middle"/>
      </xacro:unless>
      <xacro:finger_thumb_collisions  prefix="" link_prefix="rf" far="1" lf="0" />
    </xacro:if>
    <xacro:unless value="$(arg is_lite)">
      <xacro:if value="$(arg lf)">
          <xacro:if value="$(arg lf_bio)">
            <xacro:property name="distal" value="biotac"/>
            <xacro:property name="middle" value="adapter"/>
          </xacro:if>
          <xacro:unless value="$(arg lf_bio)">  
            <xacro:property name="distal" value="distal"/>
            <xacro:property name="middle" value="middle"/>
          </xacro:unless>
          <xacro:finger_thumb_collisions  prefix="" link_prefix="lf" far="1" lf="1" />
      </xacro:if>
    </xacro:unless>
  </xacro:if>
  
  <!-- pairs with other fingers for collision-->
  <xacro:if value="$(arg ff)">
    <xacro:if value="$(arg mf)">
      <xacro:if value="$(arg ff_bio)">
        <xacro:property name="distal" value="biotac"/>
        <xacro:property name="middle" value="adapter"/>
      </xacro:if>
      <xacro:unless value="$(arg ff_bio)">  
        <xacro:property name="distal" value="distal"/>
        <xacro:property name="middle" value="middle"/>
      </xacro:unless>
      <xacro:if value="$(arg mf_bio)">
        <xacro:property name="distal2" value="biotac"/>
        <xacro:property name="middle2" value="adapter"/>
        </xacro:if>
      <xacro:unless value="$(arg mf_bio)">  
        <xacro:property name="distal2" value="distal"/>
        <xacro:property name="middle2" value="middle"/>
      </xacro:unless>
      <xacro:finger_closefinger_collisions prefix="" link_prefix="ff" link_prefix2="mf"  lf="0"/>
    </xacro:if>
    <xacro:if value="$(arg rf)">
      <xacro:if value="$(arg ff_bio)">
        <xacro:property name="distal" value="biotac"/>
        <xacro:property name="middle" value="adapter"/>
      </xacro:if>
      <xacro:unless value="$(arg ff_bio)">  
        <xacro:property name="distal" value="distal"/>
        <xacro:property name="middle" value="middle"/>
      </xacro:unless>
      <xacro:if value="$(arg rf_bio)">
        <xacro:property name="distal2" value="biotac"/>
        <xacro:property name="middle2" value="adapter"/>
        </xacro:if>
      <xacro:unless value="$(arg rf_bio)">  
        <xacro:property name="distal2" value="distal"/>
        <xacro:property name="middle2" value="middle"/>
      </xacro:unless>
      <xacro:finger_farfinger_collisions prefix="" link_prefix="ff" link_prefix2="rf"  lf="0" />
    </xacro:if>   
    <xacro:unless value="$(arg is_lite)">
      <xacro:if value="$(arg lf)">
        <xacro:if value="$(arg ff_bio)">
          <xacro:property name="distal" value="biotac"/>
          <xacro:property name="middle" value="adapter"/>
        </xacro:if>
        <xacro:unless value="$(arg ff_bio)">  
          <xacro:property name="distal" value="distal"/>
          <xacro:property name="middle" value="middle"/>
        </xacro:unless>
        <xacro:if value="$(arg lf_bio)">
          <xacro:property name="distal2" value="biotac"/>
          <xacro:property name="middle2" value="adapter"/>
          </xacro:if>
        <xacro:unless value="$(arg lf_bio)">  
          <xacro:property name="distal2" value="distal"/>
          <xacro:property name="middle2" value="middle"/>
        </xacro:unless>
        <xacro:first_little_collisions prefix="" />
      </xacro:if> 
    </xacro:unless>
  </xacro:if> 

  <xacro:if value="$(arg mf)">
    <xacro:if value="$(arg rf)">
      <xacro:if value="$(arg mf_bio)">
        <xacro:property name="distal" value="biotac"/>
        <xacro:property name="middle" value="adapter"/>
      </xacro:if>
      <xacro:unless value="$(arg mf_bio)">  
        <xacro:property name="distal" value="distal"/>
        <xacro:property name="middle" value="middle"/>
      </xacro:unless>
      <xacro:if value="$(arg rf_bio)">
        <xacro:property name="distal2" value="biotac"/>
        <xacro:property name="middle2" value="adapter"/>
        </xacro:if>
      <xacro:unless value="$(arg rf_bio)">  
        <xacro:property name="distal2" value="distal"/>
        <xacro:property name="middle2" value="middle"/>
      </xacro:unless>
      <xacro:finger_closefinger_collisions prefix="" link_prefix="mf" link_prefix2="rf"  lf="0" />
    </xacro:if>
    <xacro:unless value="$(arg is_lite)">
      <xacro:if value="$(arg lf)">
        <xacro:if value="$(arg mf_bio)">
          <xacro:property name="distal" value="biotac"/>
          <xacro:property name="middle" value="adapter"/>
        </xacro:if>
        <xacro:unless value="$(arg mf_bio)">  
          <xacro:property name="distal" value="distal"/>
          <xacro:property name="middle" value="middle"/>
        </xacro:unless>
        <xacro:if value="$(arg lf_bio)">
          <xacro:property name="distal2" value="biotac"/>
          <xacro:property name="middle2" value="adapter"/>
          </xacro:if>
        <xacro:unless value="$(arg lf_bio)">  
          <xacro:property name="distal2" value="distal"/>
          <xacro:property name="middle2" value="middle"/>
        </xacro:unless>
        <xacro:finger_farfinger_collisions prefix="" link_prefix="mf" link_prefix2="lf"  lf="1" />
      </xacro:if>   
    </xacro:unless> 
  </xacro:if>

  <xacro:if value="$(arg rf)">
    <xacro:unless value="$(arg is_lite)">
      <xacro:if value="$(arg lf)">
        <xacro:if value="$(arg rf_bio)">
          <xacro:property name="distal" value="biotac"/>
          <xacro:property name="middle" value="adapter"/>
        </xacro:if>
        <xacro:unless value="$(arg rf_bio)">  
          <xacro:property name="distal" value="distal"/>
          <xacro:property name="middle" value="middle"/>
        </xacro:unless>
        <xacro:if value="$(arg lf_bio)">
          <xacro:property name="distal2" value="biotac"/>
          <xacro:property name="middle2" value="adapter"/>
          </xacro:if>
        <xacro:unless value="$(arg lf_bio)">  
          <xacro:property name="distal2" value="distal"/>
          <xacro:property name="middle2" value="middle"/>
        </xacro:unless>
        <xacro:finger_closefinger_collisions prefix="" link_prefix="rf" link_prefix2="lf"  lf="1" />
      </xacro:if>
    </xacro:unless>
  </xacro:if>
</robot>
